File: LeetCode/Medium/0518-coin-change-ii/0518-coin-change-ii.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        # dp = [0] * (amount+1)
        # dp[0] = 1
        
        # for coin in coins:
        #     for i in range(1,amount+1):
        #         if i-coin >= 0:
        #             dp[i] += dp[i-coin]
        # return dp[amount]
        memo = {}

        def dfs(remaining, start):
            if remaining == 0:
                return 1
            if remaining < 0:
                return 0
            if (remaining, start) in memo:
                return memo[(remaining, start)]

            ans = 0
            for i in range(start, len(coins)):
                ans += dfs(remaining - coins[i], i)

            memo[(remaining, start)] = ans
            return ans

        return dfs(amount, 0)

File: LeetCode/Medium/0518-coin-change-ii/0518-coin-change-ii.cpp
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        vector<int> dp(amount+1,0);
        dp[0] = 1;
        
        for(int i=0; i<coins.size(); i++){
            for(int j=1; j<amount+1; j++){
                if(j-coins[i]>=0)
                    dp[j] += dp[j-coins[i]];
            }
        }
        // for(int i=0; i<=amount;i++)
        //     cout<<dp[i]<<" ";
        return dp[amount];
    }
};